ops/testing.py

```diff
@@ -0,0 +1,392 @@
+                name: my-charm
+                requires:
+                  db:
+                    interface: pgsql
+                ''')
+        charm = harness.initialize(MyCharm)
+        # Do initial setup here
+        relation_id = harness.add_relation('db', 'postgresql')
+        # Now enable the charm to see events as the model changes
+        harness.enable_events(charm)
+        harness.add_relation_unit(relation_id, 'postgresql/0', remote_unit_data={'key': 'value'})
+        # Check that charm has properly handled the relation_joined event for postgresql/0
+        self.assertEqual(harness.charm. ...)
+
+        :param charm_meta_yaml: The YAML metadata for the charm, defining interfaces, name, etc.
+            This can be either a string or a file.
+        """
+        # TODO: jam 2020-03-05 We probably want to take config as a parameter as well, since
+        #       it would define the default values of config that the charm would see.
+        if isinstance(charm_meta_yaml, str):
+            charm_meta_yaml = dedent(charm_meta_yaml)
+        meta = charm.CharmMeta.from_yaml(charm_meta_yaml)
+        unit_name = meta.name + '/0'
+        self.unit_name = unit_name
+        self._backend = _TestingModelBackend(unit_name)
+        self._relation_id_counter = 0
+        self._model = model.Model(unit_name, meta, self._backend)
+        self._framework = framework.Framework(":memory:", "no-disk-path", meta, self._model)
+        self._charm = None
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def framework(self):
+        return self._framework
+
+    def initialize(self, charm_cls):
+        """Instantiate the Charm class provided, using the Harness's framework and charm metadata.
+             :param charm_cls: The Charm class that should be tested. If you are just testing a
+              component, you can pass in ops.charm.CharmBase.
+          :type charm_cls: CharmBase
+        """
+
+        # The Framework adds attributes to class objects for events, etc. As such, we can't re-use
+        # the original class against multiple Frameworks. So create a locally defined class
+        # and register it.
+        # TODO: jam 2020-03-16 We are looking to changes this to Instance attributes instead of
+        #       Class attributes which should clean up this ugliness. The API can stay the same
+        class TestEvents(charm_cls.on.__class__):
+            pass
+
+        TestEvents.__name__ = charm_cls.on.__class__.__name__
+
+        class TestCharm(charm_cls):
+            on = TestEvents()
+
+        # Note: jam 2020-03-01 This is so that errors in testing say MyCharm has no attribute foo,
+        # rather than TestCharm has no attribute foo.
+        TestCharm.__name__ = charm_cls.__name__
+        the_charm = TestCharm(self._framework, self._framework.meta.name)
+        self.enable_events(the_charm)
+        return the_charm
+
+    def enable_events(self, the_charm):
```

Comment:
Why do we have to provide the charm? I thought the idea was to expose the charm under harness.charm, per the sprint conversations?  The create_harness used to take the class as a parameter, and I thought we were keeping that design and just moving it to the constructor.